### DISCUSSION TITLE
Questions on Football Players Tracking

### DISCUSSION THREAD

**Discussion by lpc-eol** [roboflow/notebooks#26] [Q&A]
Hi there! Could you share some hints on the training settings for yolov5x 's best.pth? Like how did you modify the model structure of yolov5? And also, what dataset you used to train? Only the 600 images? The video is awesome and motivates me to do better! 

In fact, I did apply yolov7 and the Bytetrack to this task. To achieve better detection of football, I try to train the model of yolov7e6e on an open dataset (Soccernet Tracking Dataset) and load the pre-train weight of yolov7e6e provided by the author with an initial training setting in 100 epochs. The input resolution is 1280. I found that the mAP@.5:.95 is only 0.42, which is very low—especially the detection accuracy of football. The training progress is shown in the following image. 

![image](https://user-images.githubusercontent.com/76684848/207885290-23988a03-6235-47c2-8bc8-003b81d0e550.png)

Thank you for your assistance.

**Comment by SkalskiP** [comment]
Hi, @lpc-eol 👋🏻! I just noticed that you tried to reach out to me over email. GitHub issues or discussions are a better choice. I tend to be much more responsive here. I basically live on GH. 

I'm not sure if you've had a chance to read my [blog post](https://blog.roboflow.com/track-football-players/), but I described in a little more depth the things I only mentioned in the video.

> Like how did you modify the model structure of yolov5?

The architecture of the model is unchanged. I just trained my own model. 

> And also, what dataset you used to train? Only the 600 images?

Yes! And I tried to minimize the amount of manual work. So I pre-annotated images using the YOLOv5 COCO model. And I augmented annotated images. So, I only do annotation refinement, not full annotation.

> The video is awesome and motivates me to do better!

You made my day with that comment! 

Also, as a general comment, I found input resolution to be a key model hyper parameter when detecting small objects like football.

**Reply by lpc-eol** [reply]
Thanks again for your sharing. I read the blog and I eager to repeat your experiment of training the custom models. Could you share your optimiser parameters and how much epochs did you trained? 😀I just woken up, and  I will start working on model training!

**Reply by SkalskiP** [reply]
Hah, no worries! I trained for 100 epochs and used 1280 as image resolution. I used the default learning rate. Let me know how it went and how can I help :)

**Reply by lpc-eol** [reply]
Hi Peter, as I mentioned, I trained yolov7e6e on an open soccer dataset and made the inference yesterday. The ball could be detected even under a complex background (i.e.  https://connectpolyu-my.sharepoint.com/:f:/g/personal/19081034d_connect_polyu_hk/EorZlJLB3-JHnIwQtW74CqsBNaS5xw90nDdopP9zGd5VyQ?e=NZgJ1G). The confidence score of the ball is not that satisfactory, I am still considering another approach to improve it. Do you have any suggestions?  Thank you for your reply and share!!!!

**Reply by angineerdev** [reply]
Hello @SkalskiP @lpc-eol , I wanted to reach out because I'm facing the same issue as @lpc-eol .  Despite trying everything, I've been unable to achieve the level of ball detection performance that you've accomplished. I'd really appreciate it if you could share the exact variables (base weights, dataset version, batch size and other parameters, etc.) that you used to train your model. I hope it's not some kind of secret! 😄

By the way, here's the command I used for training my model:
`python train.py --imgsz 1280 --batch 12 --epochs 100 --data datasets/football-players-detection-2/data.yaml --weights yolov5x6.pt`

Thank you in advance for your help.

**Reply by lpc-eol** [reply]
Hi @angineeer. During my research, I found that the yolov5x6 model (6 stands for the P6 model) has poor detection performance on a small object like the ball. I recommend you try the P5 models of YOLOv5. Also, reduce the lr0 (the initial learning rate). You may try 1e^-4 at the beginning to see the model's performance on your dataset.

**Reply by angineerdev** [reply]
Hmm, I thought it would be better to use a P6 model since it's trained on 1280p inputs, as mentioned in this post https://github.com/ultralytics/yolov5/issues/2110#issuecomment-772796919 . And if I'll be training on 1280p too, shouldn't I be using the base model trained on the same resolution?

> Also, reduce the lr0 (the initial learning rate). You may try 1e^-4

I will try this 👍🏻 

> see the model's performance on your dataset

Actually  I'm attempting to replicate the steps and obtain an identical trained model as provided by @SkalskiP . So I am using the same dataset as him.

I've made several attempts, experimenting with different models, batch sizes, input sizes, and more. Although I achieved comparable results, my model noticeably lags behind his model (especially with high-speed ball movements). As I am relatively new to this field, it's possible that I might be overlooking something simple yet crucial. 😅

Anyway, thank you so much for your help.

**Reply by SkalskiP** [reply]
@lpc-eol / @angineeer I can try to help. The problem is I'm super busy. Can I help in any way at this point?

**Reply by angineerdev** [reply]
@SkalskiP Actually, yes, I still can't figure out what I am doing differently from your training process, so I am not getting the same results as the detection performance of your trained model that is shown in your video [Football Players Tracking | YOLOv5 + ByteTRACK | Google Colab | step-by-step Tutorial](https://www.youtube.com/watch?v=QCG8QMhga9k).
I don't want to take up too much of your time, but if possible, I just need to know the exact parameters and methods you used for your model training (base model, batch size, dataset version, and if you customized anything else).
I believe your upcoming video on detecting small objects with YOLO-NAS will also be very helpful, so I am eagerly looking forward to it. 😄
And also, thank you so much for providing such valuable content and assistance.

**Reply by SkalskiP** [reply]
> And also, thank you so much for providing such valuable content and assistance.

I'm glad you like it.

Do you have some examples of result videos with your current model?

**Reply by angineerdev** [reply]
@SkalskiP Here are the results for both models: https://drive.google.com/drive/folders/1cAQB16YKWsBX_VjUZ1cr4fYGmf3STN6Y
As you can see, your model experiences significantly fewer hitches and achieves better confidence scores with the ball.

**Reply by SkalskiP** [reply]
You have done an excellent job so far! I like it a lot! I assume you used my dataset. Did you use any augmentations? What is the input resolution? What is the size of the model?

**Reply by angineerdev** [reply]
> You have done an excellent job so far! I like it a lot!

@SkalskiP It is nice to hear that, thanks 😄

> I assume you used my dataset. Did you use any augmentations?

I used version 2 of the dataset you provided at the following link: https://universe.roboflow.com/roboflow-jvuqo/football-players-detection-3zvbc/dataset/2. I didn't apply any additional augmentations apart from the ones already present in version 2.

>What is the input resolution? What is the size of the model?

Both training and inference were done at 1280p. Used yolov5x6 as the base model.

To be clear, the training command I used, as mentioned in my previous comments, is as follows:
`python train.py --imgsz 1280 --batch 12 --epochs 100 --data datasets/football-players-detection-2/data.yaml --weights yolov5x6.pt`

**Reply by SkalskiP** [reply]
I'm pretty sure I used a higher resolution than `1280p`.

**Reply by angineerdev** [reply]
> Hah, no worries! I trained for 100 epochs and used 1280 as image resolution. I used the default learning rate. Let me know how it went and how can I help :)

Ah, based on this comment, I didn't attempt to go beyond 1280. I guess you were referring to a different training; my apologies! 😅
Then, let me try higher resolutions. Thank you so much for your time and help.

**Reply by SkalskiP** [reply]
I used `1280` for training, but I think I used more than that during inference. 🤔 I'm not super sure... Eh. That is why I need to start using W&B for my experiments.
